File: leetcode/array/189.py
from typing import List


class Solution:

    def rotate(self, nums: List[int], k: int) -> None:
        k = k % len(nums)
        if k == 0:
            return
        nums[:] = nums[::-1]
        nums[:k] = nums[:k][::-1]
        nums[k:] = nums[k:][::-1]


if __name__ == '__main__':
    arr = [int(x) for x in input().strip().split(' ')]
    k = int(input().strip())
    solution = Solution()
    solution.rotate(arr, k)
    print(arr)

File: leetcode/array/189.cpp
#include <iostream>

class Solution {
 public:
  void rotate(int arr[], int len, int k) {
    k = k % 10;
    // 反转: 不新增内存
    for (int i = 0; i < len / 2; ++i) {
      std::swap(arr[i], arr[len - 1 - i]);
    }
    for (int i = 0; i < k / 2; ++i) {
      std::swap(arr[i], arr[k - 1 - i]);
    }
    for (int i = k; i < k + (len - k) / 2; ++i) {
      std::swap(arr[i], arr[len - 1 - i + k]);
    }
  }
};

int main() {
  int a[7] = {1, 2, 3, 4, 5, 6, 7};

  Solution solution = Solution();
  solution.rotate(a, 7, 3);
  for (int i = 0; i < 7; ++i) {
    std::cout << a[i] << " ";
  }
  std::cout << std::endl;
}